### DISCUSSION TITLE
How and when ActionSelector works, and can we implement and use a custom ActionSelector by registering it in startup?

### DISCUSSION THREAD

**Discussion by meaning1** [dotnet/aspnetcore#27367] [General]
I tried to implement a custom ActionSelector, like below:

```
 public class CustomActionSelector : IActionSelector
    {

        public ActionDescriptor SelectBestCandidate(RouteContext context, IReadOnlyList<ActionDescriptor> candidates)
        {
           
         //some code
        }

        public IReadOnlyList<ActionDescriptor> SelectCandidates(RouteContext context)
        {
            throw new NotImplementedException();
        }
    }
}

```
And add it in Startup.cs

`services.AddSingleton<IActionSelector, CustomActionSelector>();`

But it seems not work. I'd like to know if ASP.NET Core supports a custom ActionSelector.

**Comment by davidfowl** [comment]
`IActionSelector` is currently only used for action selection in older versions of MVC (and the non-endpoint routing version). The interface still exists if you call UseMvc but isn't used in the new routing system. Documentation for the new routing system is here https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-3.1
